### ISSUE TITLE
Analysis of [Immutable] method type parameters should consider implied names

### ISSUE THREAD

**Issue by omsmith** [Brightspace/D2L.CodeStyle#677]
```csharp
class Foo {
  public static void Bar<[Immutable] T>( T value ) {}

  public static void Baz() {
    object value = null;
    Bar<object>( value );
    Bar( value ); // Analyzer doesn't catch this
  }
} 
```

The current analyzer is registered against `GenericNameSyntax`. This doesn't catch the _implied_ `Bar<object>` call. The analyzer should instead, I expect, look at `InvocationExpression`.

The same analysis path is used to restrict type parameters of named types, and so they will likely need to be split.
